### PR TITLE
#116. Implemented and tested paging for SelfContributors.

### PR DESCRIPTION
**Pull request by criske**
PR for #116

### CODE REVIEW

File: src/main/java/com/selfxdsd/storage/SelfContributors.java [hunk 1]
```diff
@@ -121,6 +154,7 @@ public Contributors ofProject(
         if (project == null) {
             return new EmptyContributors(this);
         }
+        final Page page = super.current();
```
Comment (criske) [inline src/main/java/com/selfxdsd/storage/SelfContributors.java:157]: @amihaiemil is there a way to make ofProject faster like using just sql like grouping and such?

Comment (amihaiemil) [reply src/main/java/com/selfxdsd/storage/SelfContributors.java:157]: @criske to be honest, I have no idea : )

**Review by amihaiemil — COMMENTED** [review]
@criske looks nice. I have to find some time to study how this whole paging thing is actually working...

**Comment by amihaiemil** [comment]
@rultor merge it

**Comment by amihaiemil** [comment]
@criske apparently some tests are failing. Are you sure the data you insert does not overlap or conflict, thus breaking the test logic? There is one instance of the H2 DB for all the tests.

**Comment by criske** [comment]
@amihaiemil This is weird. On my end all test are passing
`Tests run: 75, Failures: 0, Errors: 0, Skipped: 1`

File: src/main/java/com/selfxdsd/storage/SelfContributors.java (hunk 1 above)
Comment (criske) [reply src/main/java/com/selfxdsd/storage/SelfContributors.java:157]: @amihaiemil When I have time I'll try to make it work as a single sql query, just as an exercise to learn jooq :D

**Comment by criske** [comment]
@amihaiemil I've updated with your latest PR, still no problems.

**Comment by criske** [comment]
@amihaiemil Nevertheless, I think there is a  bug in ofProject() implementation (I've debugged in master ` the `returnsProjectContributors()` test). Somehow "bob" which hasn't contract with "amihaiemil/docker-java-api" appears in the interator.

**Comment by criske** [comment]
@amihaiemil nvm "bob" is added in different part in `SelfContractsITCase#addsContract()`. We really need to cleanup db after each test, or at least revert on those that are updating the db.

**Comment by amihaiemil** [comment]
@criske right, I also remember Bob, I had issues bc if him too, in one test :))

**Comment by criske** [comment]
@amihaiemil Yep, now after I've revert adding "bob" in `SelfContractsITCase#addsContract()`, I'm getting those failed 2 tests reported by rultor

**Comment by amihaiemil** [comment]
@rultor merge it

**Comment by criske** [comment]
@amihaiemil Now I can fix them :D. Interesting that rultor caught those failed tests, but I didn't. Is the db reset before each test when merging?
And I'm using `mvn clean install -Pcheckstyle,itcases` and not the IDE test runner.

**Comment by criske** [comment]
@amihaiemil Fixed, apparently I can't count :D.

**Comment by criske** [comment]
@amihaiemil Now that I'm thinking, don't merge yet... :D `registersNewContributor()` is not cleaned up and might influence `returnsProjectContributorsOfPage`

**Comment by criske** [comment]
@amihaiemil `ofProject` pagination logic was wrong afterall and somewhat tricky to do it right. Now all work as it should. It's safe to merge :D

**Comment by amihaiemil** [comment]
@criske 

> Interesting that rultor caught those failed tests, but I didn't. Is the db reset before each test when merging?

Nope, Rultor does exactly what we do: ``$mvn clean install -Pcheckstyle,itcases``. 
I also don't know why this issue was not visible before. 

Thanks for your help! :D

**Comment by amihaiemil** [comment]
@rultor merge it

**Pull request merged**